### PR TITLE
AK / LibMarkdown: Render slugified anchor tag in heading

### PR DESCRIPTION
**Pull request by tbhaxor**
image

### CODE REVIEW

File: AK/Slugify.h
Comment (AtkinsSJ) [inline AK/Slugify.h:2]: This year is 2023.

File: Tests/AK/TestSlugify.cpp
Comment (AtkinsSJ) [inline Tests/AK/TestSlugify.cpp:2]: Also 2023.

File: AK/Slugify.h
Comment (AtkinsSJ) [inline AK/Slugify.h:13]: Please don't use stdlib headers. For example, we have equivalent functions in `AK/CharacterTypes.h`.

File: AK/Slugify.h
Comment (AtkinsSJ) [inline AK/Slugify.h:22]: Don't iterate a String's bytes. String holds text as utf-8, so a code-point can be more than 1 byte. Instead, iterate `_raw.code_points()` to get each code-point as a u32.

File: AK/Slugify.h
Comment (AtkinsSJ) [inline AK/Slugify.h:17]: 1. This being an inline function in a header will make people sad. My limited understanding is it bloats the binary size, but someone like BertalanD could explain that properly.
2. `_raw` isn't a nice variable name. `raw` is fine, just don't start names with an underscore.

File: AK/Slugify.h
Comment (AtkinsSJ) [inline AK/Slugify.h:34]: Why do we construct a string and then do a replacement on it, instead of inserting the glue character when we see whitespace?

File: AK/Slugify.h
Comment (AtkinsSJ) [inline AK/Slugify.h:25]: This only works with ascii characters (and our version for just ascii would be `is_ascii_alphanumeric(ch)`). Anything non-ascii (including accented characters like `á`) will get skipped over. You probably want to replace `á` with `a` instead? Pinging @trflynn89, does LibUnicode expose a way to do that currently? I can't find anything.

File: Tests/AK/TestSlugify.cpp
Comment (AtkinsSJ) [inline Tests/AK/TestSlugify.cpp:10]: General comment: Add some tests using non-ascii characters to make sure that works, once that's supported.

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
Nice feature, but I do have some issues.

File: AK/Slugify.h
Comment (tbhaxor) [reply AK/Slugify.h:2]: Yeah, I copied it from different file. 😆

File: AK/Slugify.h
Comment (tbhaxor) [reply AK/Slugify.h:22]: I tried this. I am getting _empty_ output.

File: AK/Slugify.h
Comment (tbhaxor) [reply AK/Slugify.h:17]: Splitting implementation to cpp file makes sense here. I will do it.

File: AK/Slugify.h
Comment (tbhaxor) [reply AK/Slugify.h:34]: It was actually to remove leading and trailing glue. Implementing within loop makes sense.

File: AK/Slugify.h
Comment (tbhaxor) [reply AK/Slugify.h:22]: Oh it was UAF bug. [here](https://discord.com/channels/830522505605283862/860165510263996427/1157364726116929598)

File: Tests/AK/TestSlugify.cpp
Comment (tbhaxor) [reply Tests/AK/TestSlugify.cpp:14]: Yes added.

File: AK/Slugify.h
Comment (trflynn89) [reply AK/Slugify.h]: I'm having trouble finding a concrete spec for slugification, but my guess would be that we would want to NFD decompose the string. So something like:
```c++
auto decomposed = Unicode::normalize(input, Unicode::NormalizationForm::NFD);
```

If `input` is something like `ábc`, you will get `a\u0301bc`. You can then slugify the string, which will drop the U+0301 accent, and you're left with `abc`.

However that requires linking in LibUnicode, which we cannot do in AK. So uh, `need_for_lib_url++`.

Comment (tbhaxor) [reply AK/Slugify.h]: I actually want it to support utf-8 string, for example [check this](https://tbhaxor.com/contributions-and-support/#financial-aid-money%F0%9F%92%B0) This is from my website, the emoji on copying serialized to `%F0%9F%92%B0`, but main point is slugification supports utf8.

Comment (trflynn89) [reply AK/Slugify.h]: Decomposing a string does not mean it only supports ASCII. In fact, decomposing "💰" does not change it at all:
```c++
auto a = Unicode::normalize("💰"sv, Unicode::NormalizationForm::NFD);
dbgln("{}", a); // Prints "💰"
```

If you're interesting in reading more about normalization, check out [`String.prototype.normalize`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/normalize).

Comment (tbhaxor) [reply AK/Slugify.h]: I got your point now. 

> However, sometimes more than one code point, or sequence of code points, can represent the same abstract character.

Normalising the string would ensure it is equivalent in binary representation.

**Comment by tbhaxor** [comment]
Can we merge it now?

File: AK/Slugify.cpp
Comment (Hendiadyoin1) [inline AK/Slugify.cpp:17]: why the underscores?

Comment (tbhaxor) [reply AK/Slugify.cpp:17]: > why the underscores?

You mean for variable names?

Comment (Hendiadyoin1) [reply AK/Slugify.cpp:17]: > 2. _raw isn't a nice variable name. raw is fine, just don't start names with an underscore.

Comment (tbhaxor) [reply AK/Slugify.cpp]: Oh yes, idk how it was missed. I have fixed it in b0cc4ab17d2fc122953fcf7c4a0efb8b43cf1859

**Comment by tbhaxor** [comment]
Let me know if everything is OK and I will squash fixup.

**Comment by tbhaxor** [comment]
@AtkinsSJ @trflynn89 Ping 0x00

File: AK/Slugify.cpp
Comment (AtkinsSJ) [inline AK/Slugify.cpp:12]: Please combine these changes into the commit that added this code, (which is in this PR,) instead of separately.

Comment (tbhaxor) [reply AK/Slugify.cpp:12]: Done

File: Userland/Utilities/slugify.cpp
Comment (AtkinsSJ) [inline Userland/Utilities/slugify.cpp:33]: This will break if they pass non-ascii like `--glue "🦬"`, and act unexpectedly for multi-character strings like `--glue "hello"`.

I'd recommend making it a custom Option (`parser.add_option({ .option_name = "..." ...});`) with a check that it's a single ascii character, and making the `glue` variable a `char` instead of a `String`.

Comment (tbhaxor) [reply Userland/Utilities/slugify.cpp:33]: Ohk we can also use char. It is implemented in the ArgumentParser 

https://github.com/SerenityOS/serenity/blob/867f7da0179c6b77c729247f16e403c37b0322fa/Userland/Libraries/LibCore/ArgsParser.h#L121-L122

Comment (tbhaxor) [reply Userland/Utilities/slugify.cpp:33]: [Char is integral type](https://en.cppreference.com/w/cpp/types/is_integral). I can test it

Comment (tbhaxor) [reply Userland/Utilities/slugify.cpp:33]: I get it now what you meant. 😅

Comment (tbhaxor) [reply Userland/Utilities/slugify.cpp:33]: It is fixed now. Please https://github.com/SerenityOS/serenity/pull/21245/commits/86a1b242b23c9352537d05e4f2a2707820a927ae

**Comment by tbhaxor** [comment]
@AtkinsSJ Ping

File: Tests/AK/TestSlugify.cpp
Comment (AtkinsSJ) [inline Tests/AK/TestSlugify.cpp:12]: You can use `""_string` for all of these tests. Instead of:
```c++
MUST(String::from_utf8("Hello"sv))
```
it'd be:
```c++
"Hello"_string
```

File: Userland/Utilities/slugify.cpp
Comment (AtkinsSJ) [inline Userland/Utilities/slugify.cpp:35]: We only accept a single ascii character, so reject any input that isn't. There's also no need to convert to a DeprecatedString.

```suggestion
            if (s.length() == 1 && is_ascii_printable(s[0])) {
                glue = s[0];
                return true;
            }
            return false;
```

(And then also mention this in the help string.)

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]

File: Tests/AK/TestSlugify.cpp
Comment (tbhaxor) [reply Tests/AK/TestSlugify.cpp:12]: Oh yeah. I forgot _string literal exists. It is cleaner. I am using it now. Thanks for checking.

File: Userland/Utilities/slugify.cpp
Comment (tbhaxor) [reply Userland/Utilities/slugify.cpp]: Yeah, I thought string is not sub-scriptable. Also if it is not a valid, I am planning to use the default value `-`. So I don't think so return `false` is required.

**Comment by tbhaxor** [comment]
@AtkinsSJ I have made changes. If everything is OK, please approve it so that I can squash the fixups.

File: Userland/Utilities/slugify.cpp
Comment (AtkinsSJ) [reply Userland/Utilities/slugify.cpp]: If what the user passed in isn't valid, you should return false. Otherwise it'll look like it accepts multi-character strings but only use the first character, which is confusing. It's better for the program to just reject the input, print the help message, and the user can correct it and try again.

Comment (tbhaxor) [reply Userland/Utilities/slugify.cpp]: I see. So this is basically to make it explicit that the argument accept single character input only.

**Comment by Hendiadyoin1** [comment]
Please always squash before requesting another review from a maintainer, then they can immediately merge when they find no issues with the code, instead of telling you to squash and disappearing for two days

**Comment by tbhaxor** [comment]
> Please always squash before requesting another review from a maintainer, then they can immediately merge when they find no issues with the code, instead of telling you to squash and disappearing for two days

I used to do it, but then someone told me to keep it in different commits as it is easy to review changes. And once everything is OK, squash fixup for merge.

**Comment by tbhaxor** [comment]
Squashed the fixups and also committed the man page for `slugify` utility.

**Comment by tbhaxor** [comment]
Hey @AtkinsSJ @trflynn89 #staff ping

**Pull request merged**
